new/sdk/networkservicemesh/common/heal/client.go

```diff
@@ -0,0 +1,143 @@
+	"runtime"
+	"time"
+
+	"github.com/golang/protobuf/ptypes/empty"
+	"github.com/networkservicemesh/networkservicemesh/controlplane/api/connection"
+	"github.com/networkservicemesh/networkservicemesh/controlplane/api/networkservice"
+	"github.com/networkservicemesh/networkservicemesh/new/sdk/networkservicemesh/core/next"
+	"github.com/networkservicemesh/networkservicemesh/new/sdk/tools/extended_context"
+	"github.com/networkservicemesh/networkservicemesh/new/sdk/tools/serialize"
+	"github.com/pkg/errors"
+	"google.golang.org/grpc"
+)
+
+type healClient struct {
+	requestors        map[string]func()
+	closers           map[string]func()
+	reported          map[string]*connection.Connection
+	onHeal            []networkservice.NetworkServiceClient
+	client            connection.MonitorConnectionClient
+	eventReceiver     connection.MonitorConnection_MonitorConnectionsClient
+	updateExecutor    serialize.Executor
+	recvEventExecutor serialize.Executor
+	cancelFunc        context.CancelFunc
+}
+
+func NewClient(client connection.MonitorConnectionClient, onHeal ...networkservice.NetworkServiceClient) networkservice.NetworkServiceClient {
+	rv := &healClient{
+		onHeal:            onHeal,
+		requestors:        make(map[string]func()),
+		closers:           make(map[string]func()),
+		reported:          make(map[string]*connection.Connection),
+		client:            client,
+		updateExecutor:    serialize.NewExecutor(),
+		eventReceiver:     nil, // This is intentionally nil
+		recvEventExecutor: nil, // This is intentionally nil
+	}
+	rv.onHeal = append(rv.onHeal, rv)
+	runtime.SetFinalizer(rv, func(f *healClient) {
+		f.updateExecutor.Exec(func() {
+			if f.cancelFunc != nil {
+				f.cancelFunc()
+			}
+		})
+	})
+	return rv
+}
+
+func (f *healClient) recvEvent() {
+	if f.eventReceiver == nil {
+		// TODO handle with proper selector and cancelFunc
+		ctx, cancelFunc := context.WithCancel(context.Background())
+		f.cancelFunc = cancelFunc
+		recv, err := f.client.MonitorConnections(ctx, nil)
+		if err != nil {
+			for _, closer := range f.closers {
+				closer()
+			}
+		}
+		f.eventReceiver = recv
+	}
+	select {
+	case <-f.eventReceiver.Context().Done():
+		f.eventReceiver = nil
+	default:
+		event, err := f.eventReceiver.Recv()
+		if err != nil {
+			event = nil
+		}
+		f.updateExecutor.Exec(func() {
+			if event.GetType() == connection.ConnectionEventType_INITIAL_STATE_TRANSFER {
+				f.reported = event.GetConnections()
+			} else if event.Type == connection.ConnectionEventType_UPDATE {
+				for _, conn := range event.GetConnections() {
+					f.reported[conn.GetId()] = conn
+				}
+			} else if event.GetType() == connection.ConnectionEventType_DELETE {
+				for _, conn := range event.GetConnections() {
+					delete(f.reported, conn.GetId())
+					if f.requestors[conn.GetId()] != nil {
+						f.requestors[conn.GetId()]()
+					}
+				}
+			}
+			for id, request := range f.requestors {
+				if _, ok := f.reported[id]; !ok && f.reported[id] != nil {
```

Comment:
If `ok` equals `false` then` f.reported[id]` equals `nil`.  That whole line is not really clear to me, maybe there is some typo?

Reply:
Also correct, fixed :)